File: packages/python/elizaos/types/events.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from enum import Enum
from typing import TYPE_CHECKING, Any

from pydantic import BaseModel, Field

from elizaos.types.primitives import UUID, Content

if TYPE_CHECKING:
    pass


class EventType(str, Enum):
    # World events
    WORLD_JOINED = "WORLD_JOINED"
    WORLD_CONNECTED = "WORLD_CONNECTED"
    WORLD_LEFT = "WORLD_LEFT"

    # Entity events
    ENTITY_JOINED = "ENTITY_JOINED"
    ENTITY_LEFT = "ENTITY_LEFT"
    ENTITY_UPDATED = "ENTITY_UPDATED"

    # Room events
    ROOM_JOINED = "ROOM_JOINED"
    ROOM_LEFT = "ROOM_LEFT"

    # Message events
    MESSAGE_RECEIVED = "MESSAGE_RECEIVED"
    MESSAGE_SENT = "MESSAGE_SENT"
    MESSAGE_DELETED = "MESSAGE_DELETED"

    # Channel events
    CHANNEL_CLEARED = "CHANNEL_CLEARED"

    # Voice events
    VOICE_MESSAGE_RECEIVED = "VOICE_MESSAGE_RECEIVED"
    VOICE_MESSAGE_SENT = "VOICE_MESSAGE_SENT"

    # Interaction events
    REACTION_RECEIVED = "REACTION_RECEIVED"
    POST_GENERATED = "POST_GENERATED"
    INTERACTION_RECEIVED = "INTERACTION_RECEIVED"

    # Run events
    RUN_STARTED = "RUN_STARTED"
    RUN_ENDED = "RUN_ENDED"
    RUN_TIMEOUT = "RUN_TIMEOUT"

    # Action events
    ACTION_STARTED = "ACTION_STARTED"
    ACTION_COMPLETED = "ACTION_COMPLETED"

    # Evaluator events
    EVALUATOR_STARTED = "EVALUATOR_STARTED"
    EVALUATOR_COMPLETED = "EVALUATOR_COMPLETED"

    # Model events
    MODEL_USED = "MODEL_USED"

    # Embedding events
    EMBEDDING_GENERATION_REQUESTED = "EMBEDDING_GENERATION_REQUESTED"
    EMBEDDING_GENERATION_COMPLETED = "EMBEDDING_GENERATION_COMPLETED"
    EMBEDDING_GENERATION_FAILED = "EMBEDDING_GENERATION_FAILED"

    # Control events
    CONTROL_MESSAGE = "CONTROL_MESSAGE"


class PlatformPrefix(str, Enum):
    DISCORD = "DISCORD"
    TELEGRAM = "TELEGRAM"
    X = "X"


class EventPayload(BaseModel):
    runtime: Any = Field(..., description="Agent runtime")
    source: str = Field(..., description="Event source")
    on_complete: Callable[[], None] | None = Field(
        default=None, alias="onComplete", description="Completion callback"
    )

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class WorldPayload(EventPayload):
    world: Any = Field(..., description="World object")
    rooms: list[Any] = Field(..., description="Rooms in the world")
    entities: list[Any] = Field(..., description="Entities in the world")


class EntityPayload(EventPayload):
    entity_id: UUID = Field(..., alias="entityId", description="Entity ID")
    world_id: UUID | None = Field(default=None, alias="worldId", description="World ID")
    room_id: UUID | None = Field(default=None, alias="roomId", description="Room ID")
    metadata: dict[str, Any] | None = Field(default=None, description="Entity metadata")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class MessagePayload(EventPayload):
    message: Any = Field(..., description="Message memory")
    callback: Callable[[Content], Awaitable[list[Any]]] | None = Field(
        default=None, description="Message callback"
    )

    model_config = {"arbitrary_types_allowed": True}


class ChannelClearedPayload(EventPayload):
    room_id: UUID = Field(..., alias="roomId", description="Room ID")
    channel_id: str = Field(..., alias="channelId", description="Channel ID")
    memory_count: int = Field(..., alias="memoryCount", description="Number of memories cleared")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class InvokePayload(EventPayload):
    """Payload for events invoked without a message."""

    world_id: UUID = Field(..., alias="worldId", description="World ID")
    user_id: str = Field(..., alias="userId", description="User ID")
    room_id: UUID = Field(..., alias="roomId", description="Room ID")
    callback: Callable[[Content], Awaitable[list[Any]]] | None = Field(
        default=None, description="Callback"
    )

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class RunEventPayload(EventPayload):
    run_id: UUID = Field(..., alias="runId", description="Run ID")
    message_id: UUID = Field(..., alias="messageId", description="Message ID")
    room_id: UUID = Field(..., alias="roomId", description="Room ID")
    entity_id: UUID = Field(..., alias="entityId", description="Entity ID")
    start_time: int = Field(..., alias="startTime", description="Start time")
    status: str = Field(..., description="Run status")
    end_time: int | None = Field(default=None, alias="endTime", description="End time")
    duration: int | None = Field(default=None, description="Duration in ms")
    error: str | None = Field(default=None, description="Error message")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class ActionEventPayload(EventPayload):
    room_id: UUID = Field(..., alias="roomId", description="Room ID")
    world: UUID = Field(..., description="World ID")
    content: Content = Field(..., description="Content")
    message_id: UUID | None = Field(default=None, alias="messageId", description="Message ID")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class EvaluatorEventPayload(EventPayload):
    evaluator_id: UUID = Field(..., alias="evaluatorId", description="Evaluator ID")
    evaluator_name: str = Field(..., alias="evaluatorName", description="Evaluator name")
    start_time: int | None = Field(default=None, alias="startTime", description="Start time")
    completed: bool | None = Field(default=None, description="Whether completed")
    error: Exception | None = Field(default=None, description="Error if failed")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class ModelEventPayload(EventPayload):
    provider: str = Field(..., description="Model provider")
    type: str = Field(..., description="Model type")
    prompt: str = Field(..., description="Prompt")
    tokens: dict[str, int] | None = Field(default=None, description="Token counts")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class EmbeddingGenerationPayload(EventPayload):
    memory: Any = Field(..., description="Memory to embed")
    priority: str | None = Field(default=None, description="Priority level")
    retry_count: int | None = Field(default=None, alias="retryCount", description="Retry count")
    max_retries: int | None = Field(default=None, alias="maxRetries", description="Max retries")
    embedding: list[float] | None = Field(default=None, description="Embedding vector")
    error: Exception | str | Any | None = Field(default=None, description="Error")
    run_id: UUID | None = Field(default=None, alias="runId", description="Run ID")

    model_config = {"populate_by_name": True, "arbitrary_types_allowed": True}


class ControlMessage(BaseModel):
    type: str = Field(..., description="Control message type")
    data: dict[str, Any] | None = Field(default=None, description="Message data")

    model_config = {"extra": "allow"}


class ControlMessagePayload(EventPayload):
    message: ControlMessage = Field(..., description="Control message")


# Event handler type
EventHandler = Callable[[EventPayload], Awaitable[None]]

# Event payload map type - maps event types to their payload types
EventPayloadMap = dict[EventType, type[EventPayload]]

File: packages/typescript/src/types/events.ts
import type { HandlerCallback } from "./components";
import type { Entity, Room, World } from "./environment";
import type { Memory } from "./memory";
import type { ControlMessage } from "./messaging";
import type { ModelTypeName } from "./model";
import type { Content, JsonValue, UUID } from "./primitives";
import type { IAgentRuntime } from "./runtime";

/**
 * Standard event types across all platforms
 */
export enum EventType {
  // World events
  WORLD_JOINED = "WORLD_JOINED",
  WORLD_CONNECTED = "WORLD_CONNECTED",
  WORLD_LEFT = "WORLD_LEFT",

  // Entity events
  ENTITY_JOINED = "ENTITY_JOINED",
  ENTITY_LEFT = "ENTITY_LEFT",
  ENTITY_UPDATED = "ENTITY_UPDATED",

  // Room events
  ROOM_JOINED = "ROOM_JOINED",
  ROOM_LEFT = "ROOM_LEFT",

  // Message events
  MESSAGE_RECEIVED = "MESSAGE_RECEIVED",
  MESSAGE_SENT = "MESSAGE_SENT",
  MESSAGE_DELETED = "MESSAGE_DELETED",

  // Channel events
  CHANNEL_CLEARED = "CHANNEL_CLEARED",

  // Voice events
  VOICE_MESSAGE_RECEIVED = "VOICE_MESSAGE_RECEIVED",
  VOICE_MESSAGE_SENT = "VOICE_MESSAGE_SENT",

  // Interaction events
  REACTION_RECEIVED = "REACTION_RECEIVED",
  POST_GENERATED = "POST_GENERATED",
  INTERACTION_RECEIVED = "INTERACTION_RECEIVED",

  // Run events
  RUN_STARTED = "RUN_STARTED",
  RUN_ENDED = "RUN_ENDED",
  RUN_TIMEOUT = "RUN_TIMEOUT",

  // Action events
  ACTION_STARTED = "ACTION_STARTED",
  ACTION_COMPLETED = "ACTION_COMPLETED",

  // Evaluator events
  EVALUATOR_STARTED = "EVALUATOR_STARTED",
  EVALUATOR_COMPLETED = "EVALUATOR_COMPLETED",

  // Model events
  MODEL_USED = "MODEL_USED",

  // Embedding events
  EMBEDDING_GENERATION_REQUESTED = "EMBEDDING_GENERATION_REQUESTED",
  EMBEDDING_GENERATION_COMPLETED = "EMBEDDING_GENERATION_COMPLETED",
  EMBEDDING_GENERATION_FAILED = "EMBEDDING_GENERATION_FAILED",

  // Control events
  CONTROL_MESSAGE = "CONTROL_MESSAGE",

  // Form events
  FORM_FIELD_CONFIRMED = "FORM_FIELD_CONFIRMED",
  FORM_FIELD_CANCELLED = "FORM_FIELD_CANCELLED",
}

/**
 * Platform-specific event type prefix
 */
export enum PlatformPrefix {
  DISCORD = "DISCORD",
  TELEGRAM = "TELEGRAM",
  X = "X",
}

/**
 * Base payload interface for all events
 */
export interface EventPayload {
  runtime: IAgentRuntime;
  source?: string;
  onComplete?: () => void;
}

/**
 * Payload for world-related events
 */
export interface WorldPayload extends EventPayload {
  world: World;
  rooms: Room[];
  entities: Entity[];
}

/**
 * Payload for entity-related events
 */
export interface EntityPayload extends EventPayload {
  entityId: UUID;
  worldId?: UUID;
  roomId?: UUID;
  metadata?: {
    originalId: string;
    username: string;
    displayName?: string;
    type?: string;
  };
}

/**
 * Payload for reaction-related events
 */
export interface MessagePayload extends EventPayload {
  message: Memory;
  callback?: HandlerCallback;
}

/**
 * Payload for channel cleared events
 */
export interface ChannelClearedPayload extends EventPayload {
  roomId: UUID;
}

/**
 * Payload for events that are invoked without a message
 */
export interface InvokePayload extends EventPayload {
  worldId: UUID;
  roomId: UUID;
  userId?: UUID;
  source?: string;
  callback?: HandlerCallback;
}

/**
 * Run event payload type
 */
export interface RunEventPayload extends EventPayload {
  runId: UUID;
  messageId: UUID;
  roomId: UUID;
  entityId: UUID;
  startTime: number | bigint;
  status: "started" | "completed" | "timeout";
  endTime?: number | bigint;
  duration?: number | bigint;
  error?: string | Error;
}

/**
 * Action event payload type
 */
export interface ActionEventPayload extends EventPayload {
  roomId: UUID;
  world: UUID;
  content: Content;
  messageId?: UUID;
}

/**
 * Evaluator event payload type
 */
export interface EvaluatorEventPayload extends EventPayload {
  evaluatorId: UUID;
  evaluatorName: string;
  startTime?: number | bigint;
  completed?: boolean;
  error?: Error;
}

/**
 * Model event payload type
 */
export interface ModelEventPayload extends EventPayload {
  type: ModelTypeName;
  tokens?: {
    prompt: number;
    completion: number;
    total: number;
  };
}

/**
 * Payload for embedding generation events
 */
export interface EmbeddingGenerationPayload extends EventPayload {
  memory: Memory;
  priority?: "high" | "normal" | "low";
  embedding?: number[];
  error?: Error | string;
  runId?: UUID;
  retryCount?: number;
  maxRetries?: number;
}

/**
 * Payload for control message events
 */
export interface ControlMessagePayload extends EventPayload {
  message: ControlMessage;
}

export interface FormFieldEventPayload extends EventPayload {
  sessionId: string;
  entityId: UUID;
  field: string;
  value?: JsonValue;
  externalData?: JsonValue;
  reason?: string;
}

/**
 * Maps event types to their corresponding payload types
 */
export interface EventPayloadMap {
  [EventType.WORLD_JOINED]: WorldPayload;
  [EventType.WORLD_CONNECTED]: WorldPayload;
  [EventType.WORLD_LEFT]: WorldPayload;
  [EventType.ENTITY_JOINED]: EntityPayload;
  [EventType.ENTITY_LEFT]: EntityPayload;
  [EventType.ENTITY_UPDATED]: EntityPayload;
  [EventType.MESSAGE_RECEIVED]: MessagePayload;
  [EventType.MESSAGE_SENT]: MessagePayload;
  [EventType.MESSAGE_DELETED]: MessagePayload;
  [EventType.VOICE_MESSAGE_RECEIVED]: MessagePayload;
  [EventType.VOICE_MESSAGE_SENT]: MessagePayload;
  [EventType.CHANNEL_CLEARED]: ChannelClearedPayload;
  [EventType.REACTION_RECEIVED]: MessagePayload;
  [EventType.POST_GENERATED]: InvokePayload;
  [EventType.INTERACTION_RECEIVED]: MessagePayload;
  [EventType.RUN_STARTED]: RunEventPayload;
  [EventType.RUN_ENDED]: RunEventPayload;
  [EventType.RUN_TIMEOUT]: RunEventPayload;
  [EventType.ACTION_STARTED]: ActionEventPayload;
  [EventType.ACTION_COMPLETED]: ActionEventPayload;
  [EventType.EVALUATOR_STARTED]: EvaluatorEventPayload;
  [EventType.EVALUATOR_COMPLETED]: EvaluatorEventPayload;
  [EventType.MODEL_USED]: ModelEventPayload;
  [EventType.EMBEDDING_GENERATION_REQUESTED]: EmbeddingGenerationPayload;
  [EventType.EMBEDDING_GENERATION_COMPLETED]: EmbeddingGenerationPayload;
  [EventType.EMBEDDING_GENERATION_FAILED]: EmbeddingGenerationPayload;
  [EventType.CONTROL_MESSAGE]: ControlMessagePayload;
  [EventType.FORM_FIELD_CONFIRMED]: FormFieldEventPayload;
  [EventType.FORM_FIELD_CANCELLED]: FormFieldEventPayload;
}

/**
 * Event handler function type
 */
export type EventHandler<T extends keyof EventPayloadMap> = (
  payload: EventPayloadMap[T],
) => Promise<void>;
